api/orchestrator/orchestrator.proto

```diff
@@ -434,34 +526,119 @@ message CloudService {
 }
 
-// Requirement represents a certain requirement that needs to be fulfilled. It
-// could be a control in a certification catalog.
-message Requirement {
+message Catalog {
   string id = 1;
 
   string name = 2;
 
   string description = 3;
 
+  repeated Category categories = 4
+      [ (tagger.tags) = "gorm:\"constraint:OnDelete:CASCADE\"" ];
+}
+
+message Category {
+  string name = 1 [ (tagger.tags) = "gorm:\"primaryKey\"" ];
+
+  // Reference to the catalog this category belongs to.
+  string catalog_id = 2 [ (tagger.tags) = "gorm:\"primaryKey\"" ];
+
+  string description = 3;
+
+  repeated Control controls = 4 [
+    (tagger.tags) =
+        "gorm:\"foreignKey:category_name,category_"
+        "catalog_id;references:name,catalog_id;constraint:OnDelete:CASCADE\""
+  ];
+}
+
+// Control represents a certain Control that needs to be fulfilled. It could be
+// a Control in a certification catalog. It follows the OSCAL model. A
+// requirement in the EUCS terminology, e.g., is represented as the lowest
+// sub-control.
+message Control {
+  // A short name, e.g. OPS-01
+  // TODO(oxisto): name this "id"? or is it too confusing that this id is not
+  // unique? however, it is usually referred to as a control ID
+  string short_name = 1 [ (tagger.tags) = "gorm:\"primaryKey\"" ];
+  string category_name = 6 [ (tagger.tags) = "gorm:\"primaryKey\"" ];
+  string category_catalog_id = 7 [ (tagger.tags) = "gorm:\"primaryKey\"" ];
+
+  // Human-readable name of the control
+  string name = 2;
+
+  // Description of the control
+  string description = 3;
+
+  // List of sub - controls -
+  //     this is in accordance with the OSCAL model.
+  repeated Control controls = 4
+      [ (tagger.tags) =
+            "gorm:\"foreignKey:parent_control_short_name,parent_control_"
+            "category_name,parent_control_category_catalog_id;references=short_"
+            "name,category_name;category_catalog_id\"" ];
   // metrics contains either a list of reference to metrics - in this case only
   // the id field of the metric is populated - or a list of populated metric
   // meta-data, most likely returned by the database.
-  repeated Metric metrics = 4
-      [ (tagger.tags) = "gorm:\"many2many:requirement_metrics\"" ];
+  repeated Metric metrics = 5
+      [ (tagger.tags) =
+            "gorm:\"many2many:control_metrics;constraint:OnDelete:CASCADE\"" ];
 
-  // category can be used to categorize requirements, e.g., according to a
-  // specific security catalog or something that groups requirements together,
-  // e.g., a security control.
-  string category = 5;
+  // Reference to the parent category this control belongs to.
+
+  optional string parent_control_short_name = 8;
+  optional string parent_control_category_name = 10;
+  optional string parent_control_category_catalog_id = 9;
+}
+
+// A Target of Evaluation binds a cloud service to a catalog, so the service is
+// evaluated regarding this catalog's controls
+message TargetOfEvaluation {
```

Comment:
This could probably be modelled as a customised join table. https://gorm.io/docs/many_to_many.html#Customize-JoinTable

Should we extract this into a separate PR to think about this more careful? Maybe once our Gorm expert is back from vacation.

Reply:
I have prepared this now and it seems to work, at last for `CreateTargetOfEvaluation`. The remaining methods need to implemented and tested.